paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/BucketProcessor.scala

```diff
@@ -123,3 +131,105 @@ case class DynamicBucketProcessor(
     }
   }
 }
+
+case class GlobalDynamicBucketProcessor(
+    fileStoreTable: FileStoreTable,
+    rowType: RowType,
+    numAssigners: Integer,
+    encoderGroup: EncoderSerDeGroup)
+  extends BucketProcessor[(KeyPartOrRow, Array[Byte])] {
+
+  override def processPartition(
+      rowIterator: Iterator[(KeyPartOrRow, Array[Byte])]): Iterator[Row] = {
+    new GlobalIndexAssignerIterator(
+      rowIterator,
+      fileStoreTable,
+      rowType,
+      numAssigners,
+      encoderGroup)
+  }
+}
+
+class GlobalIndexAssignerIterator(
+    rowIterator: Iterator[(KeyPartOrRow, Array[Byte])],
+    fileStoreTable: FileStoreTable,
+    rowType: RowType,
+    numAssigners: Integer,
+    encoderGroup: EncoderSerDeGroup)
+  extends Iterator[Row]
+  with AutoCloseable {
+
+  private val queue = mutable.Queue[Row]()
+
+  val ioManager: IOManager = createIOManager
+
+  var currentResult: Row = _
+
+  var advanced = false
+
+  val assigner: GlobalIndexAssigner = {
+    val _assigner = new GlobalIndexAssigner(fileStoreTable)
+    _assigner.open(
+      0,
+      ioManager,
+      numAssigners,
+      TaskContext.getPartitionId(),
+      (row, bucket) => {
+        val extraRow: GenericRow = new GenericRow(2)
+        extraRow.setField(0, row.getRowKind.toByteValue)
+        extraRow.setField(1, bucket)
+        queue.enqueue(
+          encoderGroup.internalToRow(
+            SparkInternalRow.fromPaimon(new JoinedRow(row, extraRow), rowType)))
+      }
+    )
+    _assigner
+  }
+
+  override def hasNext: Boolean = {
+    advanceIfNeeded()
+    currentResult != null
+  }
+
+  override def next(): Row = {
+    if (!hasNext) {
+      throw new NoSuchElementException
+    }
+    advanced = false
+    currentResult
+  }
+
+  def advanceIfNeeded(): Unit = {
+    if (!advanced) {
+      advanced = true
+      currentResult = null
+      var stop = false
+      while (!stop) {
+        if (queue.nonEmpty) {
+          currentResult = queue.dequeue()
+          stop = true
+        } else if (rowIterator.hasNext) {
+          val tuple: (KeyPartOrRow, Array[Byte]) = rowIterator.next()
+          val internalRow = SerializationUtils.deserializeBinaryRow(tuple._2)
+          tuple._1 match {
+            case KeyPartOrRow.KEY_PART => assigner.bootstrapKey(internalRow)
+            case KeyPartOrRow.ROW => assigner.processInput(internalRow)
+            case _ =>
+              throw new UnsupportedOperationException(s"unknown kind ${tuple._1}")
+          }
+        } else if (assigner.inBoostrap()) {
+          assigner.endBoostrap(true)
```

Comment:
Actually, only emit records here... So your heap memory will OOM.

I refactor a lazy method for this:
https://github.com/apache/paimon/pull/3978